pkg/mimir/modules.go

```diff
@@ -576,18 +576,40 @@ func (t *Mimir) initQueryFrontend() (serv services.Service, err error) {
 	handler := transport.NewHandler(t.Cfg.Frontend.Handler, roundTripper, util_log.Logger, t.Registerer, t.ActivityTracker)
 	t.API.RegisterQueryFrontendHandler(handler, t.BuildInfoHandler)
 
+	var frontendSvc services.Service
 	if frontendV1 != nil {
 		t.API.RegisterQueryFrontend1(frontendV1)
 		t.Frontend = frontendV1
-
-		return frontendV1, nil
+		frontendSvc = frontendV1
 	} else if frontendV2 != nil {
 		t.API.RegisterQueryFrontend2(frontendV2)
-
-		return frontendV2, nil
+		frontendSvc = frontendV2
 	}
 
-	return nil, nil
+	w := services.NewFailureWatcher()
+	return services.NewBasicService(func(_ context.Context) error {
```

Comment:
[nit] Have you considered moving this service logic to a dedicated struct which wraps the frontend `services.Service` (e.g. could be called something like `FrontendGracefulShutdownService`)?

Reply:
Peter asked me explicitly to inline the service definition like this. @pstibrany WDYT?

Reply:
I advocated against that, as it looked like overkill to me tbh. Matter of preference I guess. If we wanted to use `FrontendGracefulShutdownService` elsewhere, the situation changes.

Reply:
Ok, no problem, let's keep it as is. I agree it's a matter of preference.